_source/_posts/2023-07-27-spring-graphql-react.adoc

```diff
@@ -0,0 +1,1857 @@
+[source, tsx]
+----
+import backendAPI from "./base";
+
+let requestInterceptor: number;
+let responseInterceptor: number;
+
+export const clearInterceptors = () => {
+  backendAPI.interceptors.request.eject(requestInterceptor);
+  backendAPI.interceptors.response.eject(responseInterceptor);
+};
+
+export const setInterceptors = (accessToken: String) => {
+
+  clearInterceptors();
+
+  requestInterceptor = backendAPI.interceptors.request.use(
+    // @ts-expect-error
+    function (config) {
+      return {
+        ...config,
+        headers: {
+          ...config.headers,
+          Authorization: `Bearer ${accessToken}`,
+        },
+      };
+    },
+    function (error) {
+      console.log("request interceptor error", error);
+      return Promise.reject(error);
+    }
+  );
+};
+----
+
+Add the file `src/hooks/useAccessToken.tsx` with the following content:
+
+.useAccessToken.tsx
+[source, tsx]
+----
+import { setInterceptors } from "@/services/auth";
+import { useAuth0 } from "@auth0/auth0-react";
+import { useCallback, useState } from "react";
+
+export const useAccessToken = () => {
+  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
+  const [accessToken, setAccessToken] = useState("");
+
+  const saveAccessToken = useCallback(async () => {
+    if (isAuthenticated) {
+      try {
+        const tokenValue = await getAccessTokenSilently();
+        if (accessToken !== tokenValue) {
+          setInterceptors(tokenValue);
+          setAccessToken(tokenValue);
+        }
+      } catch (err) {
+        // Inactivity timeout
+        console.log("getAccessTokenSilently error", err);
+      }
+    }
+  }, [getAccessTokenSilently, isAuthenticated, accessToken]);
+
+  return {
+    saveAccessToken,
+  };
+};
+----
+
+The hook will call Auth0 `getAccessTokenSilently` and will trigger a token refresh if the accessToken is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
+Create the `useAsyncWithToken` hook:
+
+.useAsyncWithToken.tsx
+[source, tsx]
+----
+import { useAccessToken } from "./useAccessToken";
+import { useAsync } from "react-use-custom-hooks";
+
+export const useAsyncWithToken = <T, P, E = string>(
+  asyncOperation: () => Promise<T>, deps: any[]
+) => {
+  const { saveAccessToken } = useAccessToken();
+
+  // Add saveAccessToken to the dependency list
+  const [ data, loading, error ] = useAsync(async () => {
+    await saveAccessToken();
+    return asyncOperation();
+  }, {},  deps);
+
+  return {
+    data,
+    loading,
+    error
+  };
+};
+----
+
+
+Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:
+
+.CompanyTableContainer.tsx
+[source, tsx]
+----
+  ...
+  const { data: dataList, loading: loadingList, error: errorList } = useAsyncWithToken(() => CompanyApi.getCompanyList({ page: page - 1 }), [page]);
+  const { data: dataCount } = useAsyncWithToken(() => CompanyApi.getCompanyCount(), []);
+  ...
```

Comment:
If there's no surrounding code, you can left-align all lines in this block.

Reply:
Aligned